src/virtualenv/run.py

```diff
@@ -0,0 +1,190 @@
+    verbosity.add_argument("-q", "--quiet", action="count", dest="quiet", help="decrease verbosity", default=0)
+    options, _ = parser.parse_known_args(args)
+    verbosity_value = setup_report(options.verbose, options.quiet)
+    return options, verbosity_value
+
+
+def _get_discover(parser, args, options):
+    discover_types = _collect_discovery_types()
+    discovery_parser = parser.add_argument_group("target interpreter identifier")
+    discovery_parser.add_argument(
+        "--discovery",
+        choices=list(discover_types.keys()),
+        default=next(i for i in discover_types.keys()),
+        required=False,
+        help="interpreter discovery method",
+    )
+    options, _ = parser.parse_known_args(args, namespace=options)
+    discover_class = discover_types[options.discovery]
+    discover_class.add_parser_arguments(discovery_parser)
+    options, _ = parser.parse_known_args(args, namespace=options)
+    discover = discover_class(options)
+    return discover
+
+
+def _collect_discovery_types():
+    discover_types = {e.name: e.load() for e in get_group_named("virtualenv.discovery").values()}
+    return discover_types
+
+
+def _get_creator(interpreter, parser, options):
+    creators = _collect_creators(interpreter)
+    creator_parser = parser.add_argument_group("creator options")
+    creator_parser.add_argument(
+        "--creator",
+        choices=list(creators),
+        # prefer the built-in venv if present, otherwise fallback to first defined type
+        default="venv" if "venv" in creators else next(iter(creators), None),
+        required=False,
+        help="create environment via",
+    )
+    yield
+    if options.creator not in creators:
+        raise RuntimeError("No virtualenv implementation for {}".format(interpreter))
+    creator_class = creators[options.creator]
+    creator_class.add_parser_arguments(creator_parser, interpreter)
+    yield
+    creator = creator_class(options, interpreter)
+    yield creator
+
+
+def _collect_creators(interpreter):
+    all_creators = {e.name: e.load() for e in get_group_named("virtualenv.create").values()}
+    creators = {k: v for k, v in all_creators.items() if v.supports(interpreter)}
+    return creators
+
+
+def _get_seeder(parser, options):
+    seed_parser = parser.add_argument_group("package seeder")
+    seeder_types = _collect_seeders()
+    seed_parser.add_argument(
+        "--seeder",
+        choices=list(seeder_types.keys()),
+        default="app-data",
+        required=False,
+        help="seed packages install method",
+    )
+    seed_parser.add_argument(
+        "--without-pip",
+        help="if set forces the none seeder, used for compatibility with venv",
+        action="store_true",
+        dest="without_pip",
+    )
+    yield
+    seeder_class = seeder_types["none" if options.without_pip is True else options.seeder]
+    seeder_class.add_parser_arguments(seed_parser)
+    yield
```

Comment:
ah yes I see this now -- I'd maybe write this as three functions you call in order -- the generator trick seems way too clever (from a maintainability perspective)